service/raft/raft_group0.cc

```diff
@@ -133,31 +133,43 @@ raft_group0::raft_group0(seastar::abort_source& abort_source,
     : _abort_source(abort_source), _raft_gr(raft_gr), _ms(ms), _gossiper(gs), _qp(qp), _mm(mm), _feat(feat), _sys_ks(sys_ks), _client(client)
     , _status_for_monitoring(_raft_gr.is_enabled() ? status_for_monitoring::normal : status_for_monitoring::disabled)
 {
-    init_rpc_verbs();
     register_metrics();
 }
 
+future<> raft_group0::start() {
+    return smp::invoke_on_all([this]() {
+        init_rpc_verbs();
+    });
+}
+
 void raft_group0::init_rpc_verbs() {
-    ser::group0_rpc_verbs::register_group0_peer_exchange(&_ms, [this] (const rpc::client_info&, rpc::opt_time_point, discovery::peer_list peers) {
-        return peer_exchange(std::move(peers));
+    ser::group0_rpc_verbs::register_group0_peer_exchange(&_ms.local(), [this] (const rpc::client_info&, rpc::opt_time_point, discovery::peer_list peers) {
+        return smp::submit_to(0, [this, peers = std::move(peers)] {
+            return peer_exchange(std::move(peers));
+        });
     });
 
-    ser::group0_rpc_verbs::register_group0_modify_config(&_ms, [this] (const rpc::client_info&, rpc::opt_time_point,
+    ser::group0_rpc_verbs::register_group0_modify_config(&_ms.local(), [this] (const rpc::client_info&, rpc::opt_time_point,
             raft::group_id gid, std::vector<raft::config_member> add, std::vector<raft::server_id> del) {
-        return _raft_gr.get_server(gid).modify_config(std::move(add), std::move(del));
+        return smp::submit_to(0, [this, gid, add = std::move(add), del = std::move(del)]() {
+            return _raft_gr.get_server(gid).modify_config(std::move(add), std::move(del));
+        });
     });
 
-    ser::group0_rpc_verbs::register_get_group0_upgrade_state(&_ms,
-        std::bind_front([] (raft_group0& self, const rpc::client_info&) -> future<group0_upgrade_state> {
-            co_return (co_await self._client.get_group0_upgrade_state()).second;
-        }, std::ref(*this)));
+    ser::group0_rpc_verbs::register_get_group0_upgrade_state(&_ms.local(),
+        [this] (const rpc::client_info&) -> future<group0_upgrade_state> {
+            return smp::submit_to(0, [this]() -> future<group0_upgrade_state> {
+                const auto [holder, state] = co_await _client.get_group0_upgrade_state();
+                co_return state;
+            });
+        });
 }
 
 future<> raft_group0::uninit_rpc_verbs() {
     return when_all_succeed(
-        ser::group0_rpc_verbs::unregister_group0_peer_exchange(&_ms),
-        ser::group0_rpc_verbs::unregister_group0_modify_config(&_ms),
-        ser::group0_rpc_verbs::unregister_get_group0_upgrade_state(&_ms)
+        ser::group0_rpc_verbs::unregister_group0_peer_exchange(&_ms.local()),
+        ser::group0_rpc_verbs::unregister_group0_modify_config(&_ms.local()),
+        ser::group0_rpc_verbs::unregister_get_group0_upgrade_state(&_ms.local())
```

Comment:
There is `ser::group0_rpc_verb::unregister()` that unregister all verbs in the same idl. But this is pre-existing.

Reply:
> but IMO better to separate test framework enhancement to a separate patch.

I think we've already discussed this, but I still don't feel comfortable with this approach of separating PRs into tiny patches.

Do you have specific, pragmatic reasons for doing it? With examples of situations where it pays off. Right now I see just the pain of separating these patches (fighting with git/github) and keeping them in sync, I don't see the point of it and it makes me frustrated.

> Patches should be self-contained and do one thing.

I see this in [Scylla guidelines](https://github.com/scylladb/scylladb/blob/master/docs/dev/review-checklist.md#patch-organization), but it doesn't specify what "self-contained" exactly means, nor does it give a rationale for why it matters. To me this patch is already self-contained, as it solves one specific problem and provides a test that demonstrates the problem. The code of the test and its associated helpers is just the code that needs to be written to solve this problem.

Another reason for not separating: having infra and examples of its usage in one place simplifies understanding for the reader, since he can see how parts of the infra are used and why they are important.

Reply:
> > but IMO better to separate test framework enhancement to a separate patch.
> 
> I think we've already discussed this, but I still don't feel comfortable with this approach of separating PRs into tiny patches.
> 
> Do you have specific, pragmatic reasons for doing it? With examples of situations where it pays off. Right now I see just the pain of separating these patches (fighting with git/github) and keeping them in sync, I don't see the point of it and it makes me frustrated.

Doing work is always a pin for the developer, but if it makes life easier for a reviewer it pays of. A patch that  does only one thing  is easier to review since a reviewer does not need to guess which change belongs to what fix. If a patch breaks something having  better separated changes makes bi-section easier. If it turned out that a fix needs to be backported cherry picking only minimal necessary change become possible.

If you want to look how development with "single patch per feature" work look at Cassandra where a single patch can rewrite the whole code base. Do not be like Cassandra.

Now I am not saying each patch should be no longer than X lines. If a single feature is large and isolated the patch implementing it can be large as well.

> 
> > Patches should be self-contained and do one thing.
> 
> I see this in [Scylla guidelines](https://github.com/scylladb/scylladb/blob/master/docs/dev/review-checklist.md#patch-organization), but it doesn't specify what "self-contained" exactly means, nor does it give a rationale for why it matters. To me this patch is already self-contained, as it solves one specific problem and provides a test that demonstrates the problem. The code of the test and its associated helpers is just the code that needs to be written to solve this problem.
>

It is hard to mathematically define self contained. But isn't it clear that changing a test framework has nothing to do with fixing the problem at hand? 
 
> Another reason for not separating: having infra and examples of its usage in one place simplifies understanding for the reader, since he can see how parts of the infra are used and why they are important.

Lest get this to the extreme: To linearise schema changes you need to have raft, so why not send the raft in the same single patch with the schema changes as well. According to you it *will* simplify the understanding, no?

Good explanation on infra change patch is what needed to simplify the understanding. Having an actual usage as a separate patch does not complicate anything. Tee code is there for everyone to look at.

Reply:
OK, thanks. The strongest argument for me here is to make PR easier to review, I will stick with it while fighting with git.

Reply:
Do not underestimate backporting procedure needs. Suppose we have a another bug with different smp settings and the fix has a test case as well that uses the infra you added here. Unlike this one that other bug needs to be backported, but since it builds on this PR  a person who is doing the backport needs to split this PR manually. It is more work to him. Sometime a work he cannot do without deeper understanding of the problem.

Reply:
That's understandable, but in the end, the tradeoff here is between the convenience of the PR's author and the convenience of everyone else. The increase in inconvenience for the author simply means he will have less motivation to refactor the code, make better infrastructure etc.

Reply:
There are always more developers than maintainers. Optimizations should be done to make life of maintainers easier. In the case of this PR it is trivial to split it using `git add -p`.